### PR TITLE
Fix jenkins save builds

### LINKED ISSUE #1207: extract related commit sha for jenkins build
## Description

Looks like jenkins keeps built commit sha, and it can be collected from api
image
It's valuable for user to obtain this information for analysis.

## Describe the solution you'd like
Collect this information and convert it to domain layer accordingly

## Has the Feature been Requested Before?
Related to #894

### PR DESCRIPTION
**Pull request by warren830**
# Summary

Refactor jenkins sub tasks, to be the same structure as other plugins

### Description
694303d6e4def5c1896c638f45b56c9e8d8f4255:
Instead of old method, this pr make sub tasks more clear, we can know which file do which sub task directly
Also, after several tests, createInBatches can handle empty slice errors(create an empty slice, and pass it into createInBatches)

2c93653b5db18d505a3879073b375bc601c5ecac:
store commit sha

### Does this close any open issues?
694303d6e4def5c1896c638f45b56c9e8d8f4255 will close issue #1199, 
2c93653b5db18d505a3879073b375bc601c5ecac will close #1207

### Other Information
I checked the total number of builds, the number in db equals to that in the jenkins server
auto init gitlab - 
db: 66
server: 66
auto_test - 
db: 402
server: 402
Create_license - 
db: 203
server: 203
Create_license_new - 
db: 10
server: 10
Deploy k8s single_new - 
db: 10
server: 10
Deploy k8s test -  
db: 2173
server: 2173

### CODE REVIEW

File: plugins/jenkins/tasks/jenkins_build_collect.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+			// so we need to filter it manually
+			// timestampHalfYearAgo := time.Now().AddDate(0, -6, 0).Unix() * 1000
+			// var filteredData = make([]models.JenkinsBuildProps, 0)
+			// for _, item := range builds.Builds {
+			// 	if item.Timestamp >= timestampHalfYearAgo {
+			// 		filteredData = append(filteredData, item)
+			// 	}
+			// }
+
+			var filteredData = builds.Builds
```
Comment (klesh) [inline plugins/jenkins/tasks/jenkins_build_collect.go:47]: We should rename the variable if it wasn't actually got filtered

**Review by klesh — COMMENTED** [review]
LGTM overall, some minor fixes are needed
Can we also add a small section onto README to explain the relationship between `build` and `job`? I keep forgetting which belongs to which

File: plugins/jenkins/tasks/jenkins_build_collect.go (hunk 1 above)
Comment (klesh) [reply plugins/jenkins/tasks/jenkins_build_collect.go:47]: second thought, we should filter it, I think, because once build is done, it is done, it makes no sense to update a existing build record.

Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_collect.go]: I have changed the code to filter all builds after the last build in db

File: plugins/jenkins/tasks/jenkins_build_collect.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"github.com/merico-dev/lake/logger"
+	lakeModels "github.com/merico-dev/lake/models"
+	"github.com/merico-dev/lake/plugins/jenkins/models"
+	"github.com/merico-dev/lake/utils"
+	"gorm.io/gorm/clause"
+	"time"
+)
+
+func CollectBuilds(worker *JenkinsApiClient, scheduler *utils.WorkerScheduler) error {
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_collect.go]: ok, will fix it right now

[the remark this answers is not in the document]
Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_collect.go]: > I suspect this `ctx` should be using the one got passed in at `plugin.Execute` method , so that the canceling could function?
FIxed in 6758bda560d659c038eb7a6a06dc849e19b134c6

File: plugins/jenkins/tasks/jenkins_build_collect.go (hunk 1 above)
Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_collect.go]: Also updated doc in the newest push

File: plugins/jenkins/tasks/jenkins_build_enricher.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		err = lakeModels.Db.ScanRows(cursor, build)
+
+		if build.CommitSha != "" {
+			continue
+		}
+
+		err = scheduler.Submit(func() error {
+			logger.Debug("(enrich build) Submit", build.DisplayName)
+
+			job, err := apiClient.jenkins.GetJob(ctx, build.JobName)
```
Comment (klesh) [inline plugins/jenkins/tasks/jenkins_build_enricher.go:34]: it doesn't seem right to access api inside a Enricher, an Enricher should only depends on database, not external data

File: plugins/jenkins/tasks/jenkins_build_enricher.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			if err != nil {
+				return fmt.Errorf("failed to get jenkins job: %v", err)
+			}
+			rawBuild, err := job.GetBuild(ctx, build.Number)
```
Comment (klesh) [inline plugins/jenkins/tasks/jenkins_build_enricher.go:38]: We should move this collecting operation to collecor instead

File: plugins/jenkins/tasks/jenkins_build_enricher.go (hunk 3 above)
Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_enricher.go:34]: I fixed, cause there are some other problems in this way

File: plugins/jenkins/tasks/jenkins_build_enricher.go (hunk 4 above)
Comment (warren830) [reply plugins/jenkins/tasks/jenkins_build_enricher.go]: Sorry, already moved

**Review by klesh — APPROVED** [review]
LGTM
